raft/consensus_module.nim

```diff
@@ -5,4 +5,95 @@
+    var cnt = 0
+    for peer in node.peers:
+      if peer.hasVoted:
+        cnt.inc
+    if cnt >= (node.peers.len div 2 + node.peers.len mod 2):
+      result = true
+
+proc RaftNodeHandleHeartBeat*[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType], msg: RaftMessageAppendEntries): RaftMessageAppendEntriesResponse[SmStateType] =
+  debug "Received heart-beat", node_id=node.id, sender_id=msg.sender_id, node_current_term=node.currentTerm, sender_term=msg.senderTerm
+  result = RaftMessageAppendEntriesResponse[SmStateType](op: rmoAppendLogEntry, senderId: node.id, receiverId: msg.senderId, msgId: msg.msgId, success: false)
+  withRLock(node.raftStateMutex):
+    if msg.senderTerm >= node.currentTerm:
+      RaftNodeCancelAllTimers(node)
+      if node.state == rnsCandidate:
+        RaftNodeAbortElection(node)
+      result.success = true
+      node.currentTerm = msg.senderTerm
+      node.votedFor = DefaultUUID
+      node.currentLeaderId = msg.senderId
+      RaftNodeScheduleElectionTimeout(node)
+
+proc RaftNodeHandleRequestVote*[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType], msg: RaftMessageRequestVote): RaftMessageRequestVoteResponse =
+  withRLock(node.raftStateMutex):
+    result = RaftMessageRequestVoteResponse(msgId: msg.msgId, senderId: node.id, receiverId: msg.senderId, granted: false)
+    if node.state != rnsCandidate and node.state != rnsStopped and msg.senderTerm > node.currentTerm and node.votedFor == DefaultUUID:
+      if msg.lastLogTerm >= RaftNodeLogEntryGet(node, RaftNodeLogIndexGet(node)).term or
+        (msg.lastLogTerm == RaftNodeLogEntryGet(node, RaftNodeLogIndexGet(node)).term and msg.lastLogIndex >= RaftNodeLogIndexGet(node)):
+        asyncSpawn cancelAndWait(node.electionTimeoutTimer)
+        node.votedFor = msg.senderId
+        result.granted = true
+        RaftNodeScheduleElectionTimeout(node)
+
+proc RaftNodeAbortElection*[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType]) =
+  withRLock(node.raftStateMutex):
+    node.state = rnsFollower
+    for fut in node.votesFuts:
+        waitFor cancelAndWait(fut)
+
+proc RaftNodeStartElection*[SmCommandType, SmStateType](node: RaftNode[SmCommandType, SmStateType]) {.async.} =
+  withRLock(node.raftStateMutex):
+    node.currentTerm.inc
+    node.state = rnsCandidate
+    node.votedFor = node.id
+    debug "Raft Node started election", node_id=node.id, state=node.state, voted_for=node.votedFor
+
+    for peer in node.peers:
+      peer.hasVoted = false
+      node.votesFuts.add(node.msgSendCallback(
+          RaftMessageRequestVote(
+            op: rmoRequestVote, msgId: genUUID(), senderId: node.id,
+            receiverId: peer.id, lastLogTerm: RaftNodeLogEntryGet(node, RaftNodeLogIndexGet(node)).term,
+            lastLogIndex: RaftNodeLogIndexGet(node), senderTerm: node.currentTerm)
+        )
+      )
+
+  # Process votes (if any)
+  for voteFut in node.votesFuts:
+    let r = await voteFut
```

Comment:
Is it important to process the votes in certain order (e.g. order of arrival)? The code here will process them in order of sending and if the first vote doesn't arrive, none of the votes will be processed. Also, could we wait here forever? Is there a timeout being applied anywhere?